HW0/P4/P4.py

```diff
@@ -40,38 +41,68 @@
     # Read the observation array and plot it (Part 2)
     #####################
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '.g', label='Observed trajectory')
+    m = np.loadtxt('P4_measurements.txt',delimiter=',') #loaded measurement data using np.loadtxt like above
+    m=m.transpose() #need to transpose matrix m to align with the mathematical formula / matrix multiiplication 
+    r = np.array([[1/rx,0,0],[0,1/ry,0],[0,0,1/rz]]) #initialize r
+    x=np.dot(r,m) #needed to perform dot multiplication to account for broadcasting error
+    ax.plot(x[0,:],x[1,:],x[2,:],'.g', label='Observed trajectory')  #plotted x which is the plot that takes into account dimensional stretching
 
     #####################
     # Part 3:
     # Use the initial conditions and propagation matrix for prediction
     #####################
-
-    # A = ?
-    # a = ?
-    # s = ?
-
-    # Initial conditions for s0
+    A = np.array([[1,0,0,dt,0,0],[0,1,0,0,dt,0],[0,0,1,0,0,dt],[0,0,0,1-c*dt,0,0],[0,0,0,0,1-c*dt,0,],[0,0,0,0,0,1-c*dt]]) #initialize A array
+    a = np.array([0,0,0,0,0,g*dt]) #initialize a array
+    a = a.transpose() #transpose a
+    s = np.zeros([6,K]) #initialize s vector
+    s0=np.array([0.0,0.0,2.0,15.0,3.5,4.0]) #set s0
+    s0=s0.transpose() #tranpose s0 to align into structure of how sk will be built
+    i = 1
+    s[:,0]=s0 #assign first column 0 of sk to be equal to s0
+    
     # Compute the rest of sk using Eq (1)
-
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-k', label='Blind trajectory')
+    
+    while i <= K-1:
+        s[:,i]=np.add(np.dot(A,s[:,i-1]),a) #use np.add and np.dot like above to account for having arrays here
+        i+=1
+    
+    ax.plot(s[0,:], s[1,:], s[2,:],'-k', label='Blind trajectory') #plot blind trajectory
 
     #####################
     # Part 4:
     # Use the Kalman filter for prediction
     #####################
-
-    # B = ?
-    # C = ?
-
-    # Initial conditions for s0 and Sigma0
+    
+    def predictS(A,s,a): #to be called s_predicted in loop or sk
+        return np.add(np.dot(A,s),a) 
+                         
+    def predictSig(A,Sigma,B): #to be called Sig_predicted in loop or Sigk
+        return inv(np.add(np.dot(np.dot(A,Sigma),A.transpose()),np.dot(B,B.transpose())))
+        
+    def updateSig(x,y): #to be called Sig_updated in loop or Sig K+1
+        return inv(np.add(x,np.dot(y.transpose(),y))) 
+    
+    def updateS(x,y,z,i,j): #to be called s_new in loop or  Sk+1
+        return np.dot(x,np.add(np.dot(y,z),np.dot(i.transpose(),j))) 
+    
```

Comment:
If you want to avoid writing out the operations like .add() and .transpose() to make it clearer and to use less parentheses you can convert to matrices and use the mathematical symbols like + and *.  You can also transpose using .T and invert using .I